### PR TITLE
Ranking

### PR DESCRIPTION
**Pull request by masnn0**
:)

### CODE REVIEW

File: vj4/ui/templates/partials/nav.html [hunk 1]
```diff
@@ -31,6 +31,7 @@
     {% if handler.has_perm(vj4.model.builtin.PERM_EDIT_DESCRIPTION) or handler.has_perm(vj4.model.builtin.PERM_EDIT_PERM) %}
       {{ nav_item('domain_manage', 'domain_manage') }}
     {% endif %}
+    {{ nav_item('domain_ranking', 'domain_ranking') }}
```
Comment (twd2) [inline vj4/ui/templates/partials/nav.html]: Need discussion: where should be the ranking item placed.

File: vj4/handler/user.py [hunk 2]
```diff
@@ -253,3 +254,35 @@ def modify_udoc(self, udict, key):
     for i in range(len(udocs)):
       self.modify_udoc(udocs, i)
     self.json(udocs)
+
+
+async def render_or_json_rank(self, page, ppcount, pcount, pdocs, **kwargs):
```
Comment (twd2) [inline vj4/handler/user.py]: Why do you create a new function for this?

File: vj4/handler/user.py [hunk 3]
```diff
@@ -253,3 +254,35 @@ def modify_udoc(self, udict, key):
...
+  if 'page_title' not in kwargs:
+    kwargs['page_title'] = self.translate(self.TITLE)
+  if 'path_components' not in kwargs:
+    kwargs['path_components'] = self.build_path((self.translate(self.NAME), None))
+  if self.prefer_json:
+    list_html = self.render_html('partials/rank.html', page=page, ppcount=ppcount,
+                                 pcount=pcount, pdocs=pdocs, psdict=psdict)
+    path_html = self.render_html('partials/path.html', path_components=kwargs['path_components'])
+    self.json({'title': self.render_title(kwargs['page_title']),
```
Comment (twd2) [inline vj4/handler/user.py]: Please do not copy the code from problem list's handler and please refer to other lists.

The problem list's handler has something special.

File: vj4/handler/user.py [hunk 4]
```diff
@@ -253,3 +254,35 @@ def modify_udoc(self, udict, key):
...
+class RankHandler(base.Handler):
+  USERS_PER_PAGE = 100
+  @base.get_argument
+  @base.sanitize
+  async def get(self, *, domain_id: str="system", page: int=1):
+    pdocs, ppcount, pcount = await pagination.paginate(domain.get_multi_user(domain_id=domain_id).sort([('rp', -1)]),
+                                                       page, self.USERS_PER_PAGE)
+    for pdoc in pdocs:
+      pdoc['uname'] = await user.get_by_uid(pdoc['uid'])
+      pdoc['uname'] = pdoc['uname']['uname']
```
Comment (twd2) [inline vj4/handler/user.py]: What is this?

File: vj4/handler/user.py [hunk 5]
```diff
@@ -253,3 +254,35 @@ def modify_udoc(self, udict, key):
...
+@app.route('/ranking', 'domain_ranking')
+class RankHandler(base.Handler):
+  USERS_PER_PAGE = 100
+  @base.get_argument
+  @base.sanitize
+  async def get(self, *, domain_id: str="system", page: int=1):
+    pdocs, ppcount, pcount = await pagination.paginate(domain.get_multi_user(domain_id=domain_id).sort([('rp', -1)]),
+                                                       page, self.USERS_PER_PAGE)
+    for pdoc in pdocs:
+      pdoc['uname'] = await user.get_by_uid(pdoc['uid'])
```
Comment (twd2) [inline vj4/handler/user.py]: Use `user.get_dict`

File: vj4/handler/user.py [hunk 6]
```diff
@@ -253,3 +254,35 @@ def modify_udoc(self, udict, key):
...
+    self.render('ranking.html', page=page, ppcount=ppcount, pcount=pcount, pdocs=pdocs, **kwargs)
+
+
+@app.route('/ranking', 'domain_ranking')
+class RankHandler(base.Handler):
+  USERS_PER_PAGE = 100
+  @base.get_argument
+  @base.sanitize
+  async def get(self, *, domain_id: str="system", page: int=1):
+    pdocs, ppcount, pcount = await pagination.paginate(domain.get_multi_user(domain_id=domain_id).sort([('rp', -1)]),
```
Comment (twd2) [inline vj4/handler/user.py]: `udocs`

File: vj4/ui/templates/partials/ranking.html [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        <th class="col--rp">{{ _('RP') }}</th>
+        <th class="col--ac">{{ _('Accepted') }}</th>
+      </tr>
+    </thead>
+    <tbody>
+    {% for pdoc in pdocs %}
+      {% if pdoc['rp']>0 %}
+        <tr>
+          <td class="col--name">{{ pdoc['uname'] }}</td>
+          <td class="col--rp">{{ pdoc['rp']|int }}</td>
```
Comment (twd2) [inline vj4/ui/templates/partials/ranking.html]: Why int?

File: vj4/ui/templates/ranking.html [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{% extends "layout/basic.html" %}
+{% block content %}
+<div class="row">
+  <div class="medium-12 columns">
+    <div class="section">
+      <div class="section__body no-padding">
+        {% include "partials/ranking.html" %}
```
Comment (twd2) [inline vj4/ui/templates/ranking.html]: You don't have to create a partial page for this.

File: vj4/ui/templates/partials/ranking.html (hunk 7 above)
Comment (masnn) [reply vj4/ui/templates/partials/ranking.html]: Or you'll see a long float number.

File: vj4/ui/templates/ranking.html (hunk 8 above)
Comment (masnn) [reply vj4/ui/templates/ranking.html]: It's modified from problem_list,I'll fix it later.

File: vj4/handler/user.py (hunk 4 above)
Comment (masnn) [reply vj4/handler/user.py]: Table `user` contains uname&uid,
But table `domain.user` contains uid&rp.
We have to put then together.

File: vj4/handler/user.py [hunk 9]
```diff
@@ -13,6 +13,7 @@
 from vj4.model.adaptor import discussion
 from vj4.model.adaptor import problem
 from vj4.model.adaptor import setting
+from vj4.util import pagination
 from vj4.util import misc
 from vj4.util import options
```
Comment (twd2) [inline vj4/handler/user.py]: Please put `from vj4.util import pagination` after this line.

File: vj4/handler/user.py [hunk 10]
```diff
@@ -253,3 +254,17 @@ def modify_udoc(self, udict, key):
...
       self.modify_udoc(udocs, i)
     self.json(udocs)
+
+
+@app.route('/ranking', 'domain_ranking')
+class RankHandler(base.Handler):
+  USERS_PER_PAGE = 100
+  @base.get_argument
+  @base.sanitize
+  async def get(self, *, domain_id: str="system", page: int=1):
```
Comment (twd2) [inline vj4/handler/user.py]: The parameter `domain_id` is not needed, please use `self.domain_id`.

File: vj4/handler/user.py [hunk 11]
```diff
@@ -253,3 +254,17 @@ def modify_udoc(self, udict, key):
     for i in range(len(udocs)):
       self.modify_udoc(udocs, i)
     self.json(udocs)
+
+
+@app.route('/ranking', 'domain_ranking')
+class RankHandler(base.Handler):
+  USERS_PER_PAGE = 100
+  @base.get_argument
```
Comment (twd2) [inline vj4/handler/user.py]: An empty line, please.

File: vj4/handler/user.py [hunk 12]
```diff
@@ -253,3 +254,17 @@ def modify_udoc(self, udict, key):
...
+  @base.sanitize
+  async def get(self, *, domain_id: str="system", page: int=1):
+    udocs, uucount, ucount = await pagination.paginate(domain.get_multi_user(domain_id=domain_id).sort([('rp', -1)]),
```
Comment (twd2) [inline vj4/handler/user.py]: Use `self.domain_id` here.

Comment (twd2) [reply vj4/handler/user.py]: Considering using the `rank` field.

File: vj4/handler/user.py [hunk 13]
```diff
@@ -253,3 +254,17 @@ def modify_udoc(self, udict, key):
...
+                                                       page, self.USERS_PER_PAGE)
+    for udoc in udocs:
+      udoc['uname'] = await user.get_by_uid(udoc['uid'])
+      udoc['uname'] = udoc['uname']['uname']
```
Comment (twd2) [inline vj4/handler/user.py]: Looks too strange.

File: vj4/handler/user.py [hunk 14]
```diff
@@ -253,3 +254,17 @@ def modify_udoc(self, udict, key):
...
+@app.route('/ranking', 'domain_ranking')
+class RankHandler(base.Handler):
+  USERS_PER_PAGE = 100
+  @base.get_argument
+  @base.sanitize
+  async def get(self, *, domain_id: str="system", page: int=1):
+    udocs, uucount, ucount = await pagination.paginate(domain.get_multi_user(domain_id=domain_id).sort([('rp', -1)]),
+                                                       page, self.USERS_PER_PAGE)
+    for udoc in udocs:
+      udoc['uname'] = await user.get_by_uid(udoc['uid'])
```
Comment (twd2) [inline vj4/handler/user.py]: Use this: https://github.com/vijos/vj4/blob/master/vj4/model/user.py#L94

File: vj4/ui/templates/ranking.html [hunk 15]
```diff
@@ -0,0 +1,44 @@
+{% extends "layout/basic.html" %}
+{% import "components/paginator.html" as paginator with context %}
+{% import "components/nothing.html" as nothing with context %}
+{% block content %}
+<div class="row">
+	<div class="medium-12 columns">
```
Comment (twd2) [inline vj4/ui/templates/ranking.html]: No TAB.

Comment (twd2) [reply vj4/ui/templates/ranking.html]: The indentations are completely wrong.

File: vj4/ui/templates/ranking.html [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+					{% if not udocs %}
+					{{ nothing.render('Sorry, there is no user in this domain.') }}
+					{% else %}
+					<table class="data-table">
+						<colgroup>
+							<col class="col--name">
+							<col class="col--rp">
+						</colgroup>
+						<thead>
+							<tr>
```
Comment (twd2) [inline vj4/ui/templates/ranking.html]: It's better to add a column for the field `rank`.

File: vj4/ui/templates/ranking.html [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+	<div class="medium-12 columns">
+		<div class="section">
+			<div class="section__body no-padding">
+				<div data-fragment-id="rank">
+					{% if not udocs %}
+					{{ nothing.render('Sorry, there is no user in this domain.') }}
+					{% else %}
+					<table class="data-table">
+						<colgroup>
+							<col class="col--name">
```
Comment (twd2) [inline vj4/ui/templates/ranking.html]: Please make this consistent with below.

File: vj4/ui/templates/ranking.html [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+						<thead>
+							<tr>
+								<th class="col--name">{{ _('Username') }}</th>
+								<th class="col--rp">{{ _('RP') }}</th>
+								<th class="col--ac">{{ _('Accepted') }}</th>
+							</tr>
+						</thead>
+						<tbody>
+							{% for udoc in udocs %}
+							{% if udoc['rp']>0 %}
```
Comment (twd2) [inline vj4/ui/templates/ranking.html]:
```
] > 0
 ^ ^
```

File: vj4/ui/templates/ranking.html [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+							<tr>
+								<td class="col--name">{{ udoc['uname'] }}</td>
+								<td class="col--rp">{{ udoc['rp']|int }}</td>
```
Comment (twd2) [inline vj4/ui/templates/ranking.html]: See: https://github.com/vijos/vj4/blob/master/vj4/ui/templates/user_detail.html#L32

File: vj4/ui/templates/ranking.html [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+								<th class="col--name">{{ _('Username') }}</th>
+								<th class="col--rp">{{ _('RP') }}</th>
+								<th class="col--ac">{{ _('Accepted') }}</th>
+							</tr>
+						</thead>
+						<tbody>
+							{% for udoc in udocs %}
+							{% if udoc['rp']>0 %}
+							<tr>
+								<td class="col--name">{{ udoc['uname'] }}</td>
```
Comment (twd2) [inline vj4/ui/templates/ranking.html]: Use `user.render_inline`, see: https://github.com/vijos/vj4/blob/master/vj4/ui/templates/contest_scoreboard.html#L39

File: vj4/ui/templates/partials/nav.html (hunk 1 above)
Comment (twd2) [reply vj4/ui/templates/partials/nav.html]: Soha says footer, while I don't agree.

File: vj4/handler/user.py (hunk 13 above)
Comment (masnn) [reply vj4/handler/user.py]: Please give a better solution.

File: vj4/handler/user.py (hunk 4 above)
Comment (twd2) [reply vj4/handler/user.py]: Please note that we should not modify the docs returned from models' methods. This is the convention.

Use `dudocs` and `udict` like other handlers.

**Comment by twd2** [comment]
> maybe `udict[udoc['_id']]` (?

Yes. I was wrong.

**Comment by twd2** [comment]
> > maybe `udict[udoc['_id']]` (?
> 
> Yes. I was wrong.

I was correct. You were wrong.

**Comment by twd2** [comment]
Please check https://github.com/masnn0/vj4/pull/6

**Comment by twd2** [comment]
image

**Pull request merged**